From c452da86ea73d04b66a75e841c87e95e24d5ecb3 Mon Sep 17 00:00:00 2001
From: Daniel Nephin <dnephin@gmail.com>
Date: Mon, 24 Jun 2019 20:10:32 -0400
Subject: [PATCH] Update tests for latest google/go-cmp

Change the expected value for a couple of them to be less strict

Backport to v2.3.0.
https://github.com/gotestyourself/gotest.tools/commit/85ad333
---
 assert/assert_test.go      | 10 +++-------
 assert/cmp/compare_test.go | 37 +++++++++++++++++++------------------
 assert/opt/opt_test.go     | 23 +++++++++++++++++++++--
 3 files changed, 43 insertions(+), 27 deletions(-)

diff --git a/assert/assert_test.go b/assert/assert_test.go
index 61d2853..4d117ea 100644
--- a/assert/assert_test.go
+++ b/assert/assert_test.go
@@ -302,13 +302,9 @@ func TestDeepEqualFailure(t *testing.T) {
 
 	fakeT := &fakeTestingT{}
 	DeepEqual(fakeT, actual, expected, gocmp.AllowUnexported(stub{}))
-	expectFailNowed(t, fakeT, "assertion failed: "+`
---- actual
-+++ expected
-{assert.stub}.b:
-	-: 1
-	+: 2
-`)
+	if !fakeT.failNowed {
+		t.Fatal("should have failNowed")
+	}
 }
 
 func TestErrorFailure(t *testing.T) {
diff --git a/assert/cmp/compare_test.go b/assert/cmp/compare_test.go
index f99ce88..7878b0b 100644
--- a/assert/cmp/compare_test.go
+++ b/assert/cmp/compare_test.go
@@ -14,32 +14,33 @@ import (
 )
 
 func TestDeepEqual(t *testing.T) {
-	actual := DeepEqual([]string{"a", "b"}, []string{"b", "a"})()
-	expected := `
---- result
-+++ exp
-{[]string}[0]:
-	-: "a"
-	+: "b"
-{[]string}[1]:
-	-: "b"
-	+: "a"
-`
-	args := []ast.Expr{&ast.Ident{Name: "result"}, &ast.Ident{Name: "exp"}}
-	assertFailureTemplate(t, actual, args, expected)
+	t.Run("failure", func(t *testing.T) {
+		result := DeepEqual([]string{"a", "b"}, []string{"b", "a"})()
+		if result.Success() {
+			t.Errorf("expected failure")
+		}
+
+		args := []ast.Expr{&ast.Ident{Name: "result"}, &ast.Ident{Name: "exp"}}
+		message := result.(templatedResult).FailureMessage(args)
+		expected := "\n--- result\n+++ exp\n"
+		if !strings.HasPrefix(message, expected) {
+			t.Errorf("expected prefix \n%q\ngot\n%q\n", expected, message)
+		}
+	})
 
-	actual = DeepEqual([]string{"a"}, []string{"a"})()
-	assertSuccess(t, actual)
+	t.Run("success", func(t *testing.T) {
+		actual := DeepEqual([]string{"a"}, []string{"a"})()
+		assertSuccess(t, actual)
+	})
 }
 
 type Stub struct {
 	unx int
 }
 
-func TestDeepEqualeWithUnexported(t *testing.T) {
+func TestDeepEqualWithUnexported(t *testing.T) {
 	result := DeepEqual(Stub{}, Stub{unx: 1})()
-	assertFailure(t, result, `cannot handle unexported field: {cmp.Stub}.unx
-consider using AllowUnexported or cmpopts.IgnoreUnexported`)
+	assertFailureHasPrefix(t, result, `cannot handle unexported field: {cmp.Stub}.unx`)
 }
 
 func TestRegexp(t *testing.T) {
diff --git a/assert/opt/opt_test.go b/assert/opt/opt_test.go
index 734ab71..b5bfb77 100644
--- a/assert/opt/opt_test.go
+++ b/assert/opt/opt_test.go
@@ -186,7 +186,12 @@ func TestPathStringFromStruct(t *testing.T) {
 
 	spec := "Ref.Children.Labels.Value"
 	matches := matchPaths(fixture, PathString(spec))
-	expected := []string{`{opt.node}.Ref.Children[1].Labels["first"].Value`}
+	expected := []string{
+		`{opt.node}.Ref.Children[1].Labels["first"].Value`,
+		// as of google/go-cmp 0.3.0 PathFilter seems to traverse some parts
+		// of the tree more than once.
+		`{opt.node}.Ref.Children[1].Labels["first"].Value`,
+	}
 	assert.DeepEqual(t, matches, expected)
 }
 
@@ -211,7 +216,14 @@ func TestPathStringFromSlice(t *testing.T) {
 
 	spec := "Ref.Children.Labels.Ref.Value"
 	matches := matchPaths(fixture, PathString(spec))
-	expected := []string{`{[]opt.node}[0].Ref.Children[1].Labels["second"].Ref.Value`}
+	expected := []string{
+		`{[]opt.node}[0].Ref.Children[1].Labels["second"].Ref.Value`,
+		// as of google/go-cmp 0.3.0 PathFilter seems to traverse some parts
+		// of the tree more than once.
+		`{[]opt.node}[0].Ref.Children[1].Labels["second"].Ref.Value`,
+		`{[]opt.node}[0].Ref.Children[1].Labels["second"].Ref.Value`,
+		`{[]opt.node}[0].Ref.Children[1].Labels["second"].Ref.Value`,
+	}
 	assert.DeepEqual(t, matches, expected)
 }
 
@@ -233,6 +245,13 @@ func TestPathField(t *testing.T) {
 		"{opt.node}.Children[1].Value.Value",
 		"{opt.node}.Children[2].Value.Value",
 		"{opt.node}.Children[2].Ref.Value.Value",
+
+		// as of google/go-cmp 0.3.0 PathFilter seems to traverse some parts
+		// of the tree twice.
+		"{opt.node}.Children[0].Value.Value",
+		"{opt.node}.Children[1].Value.Value",
+		"{opt.node}.Children[2].Value.Value",
+		"{opt.node}.Children[2].Ref.Value.Value",
 	}
 	assert.DeepEqual(t, matches, expected)
 }
